src/main/kotlin/cacophony/semantic/VarUseTypeAnalysis.kt

```diff
@@ -11,5 +9,227 @@ fun analyzeVarUseTypes(
+                    expr,
+                    UseTypesForExpression.merge(
+                        useTypeAnalysis.get(expr.lhs),
+                        useTypeAnalysis.get(expr.rhs),
+                    ),
+                )
+            }
+        }
+    }
+
+    private fun visitCompoundAssignment(expr: OperatorBinary) {
+        visitExpression(expr.rhs)
+        when (expr.lhs) {
+            is VariableUse -> {
+                useTypeAnalysis.put(expr, UseTypesForExpression.empty())
+                useTypeAnalysis.get(expr)!!.add(
+                    resolvedVariables[expr.lhs]!!,
+                    VariableUseType.READ_WRITE,
+                )
+                useTypeAnalysis.get(expr)!!.mergeWith(useTypeAnalysis.get(expr.rhs)!!)
+            }
+
+            else -> visitExpression(expr.lhs)
+        }
+    }
+
+    private fun visitAssignment(expr: OperatorBinary.Assignment) {
+        visitExpression(expr.rhs)
+        when (expr.lhs) {
+            is VariableUse -> {
+                useTypeAnalysis.put(expr, UseTypesForExpression.empty())
+                useTypeAnalysis.get(expr)!!.add(
+                    resolvedVariables[expr.lhs]!!,
+                    VariableUseType.WRITE,
+                )
+                useTypeAnalysis.get(expr)!!.mergeWith(useTypeAnalysis.get(expr.rhs)!!)
+            }
+
+            else -> visitExpression(expr.lhs)
+        }
+    }
+
+    private fun visitUnaryOperator(expr: OperatorUnary) {
+        visitExpression(expr.expression)
+        useTypeAnalysis.put(
+            expr,
+            UseTypesForExpression.merge(
+                useTypeAnalysis.get(expr.expression),
+            ),
+        )
+    }
+
+    private fun visitReturnStatement(expr: Statement.ReturnStatement) {
+        visitExpression(expr.value)
+        useTypeAnalysis.put(
+            expr,
+            UseTypesForExpression.merge(
+                useTypeAnalysis.get(expr.value),
+            ),
+        )
+    }
+
+    private fun visitWhileStatement(expr: Statement.WhileStatement) {
+        visitExpression(expr.testExpression)
+        visitExpression(expr.doExpression)
+        useTypeAnalysis.put(
+            expr,
+            UseTypesForExpression.merge(
+                useTypeAnalysis.get(expr.testExpression),
+                useTypeAnalysis.get(expr.doExpression),
+            ),
+        )
+    }
+
+    private fun visitIfElseStatement(expr: Statement.IfElseStatement) {
+        visitExpression(expr.testExpression)
+        visitExpression(expr.doExpression)
+        expr.elseExpression?.let { visitExpression(it) }
+        useTypeAnalysis.put(
+            expr,
+            UseTypesForExpression.merge(
+                useTypeAnalysis.get(expr.testExpression),
+                useTypeAnalysis.get(expr.doExpression),
+                useTypeAnalysis.get(expr.elseExpression),
+            ),
+        )
+    }
+
+    private fun visitFunctionCall(expr: FunctionCall) {
```

Comment:
I think here you have to check how this function uses variables (probably via `analyzedFunction`, so ad. 3. you might need this, instead of `callGraph`, sorry). e.g. in `x+f[]`, where `f` uses `x` , `x` should be marked as used in `f[]`, so later the control flow is correct.